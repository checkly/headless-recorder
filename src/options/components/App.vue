<template>
  <div class="options">
    <div class="container">
      <div class="header">
        Playwright Recorder Options
        <small class="saving-badge text-muted" v-show="saving">
          Saving...
        </small>
      </div>
      <div class="content" v-if="!loading">
        <div class="settings-block">
          <h4 class="settings-block-title">
            Code Recorder settings
          </h4>
          <div class="settings-block-main">
            <div class="settings-group">
              <label class="settings-label">custom data attribute</label>
              <input id="options-code-dataAttribute" type="text" v-model="options.code.dataAttribute" @change="save" placeholder="your custom data-* attribute">
              <small>Define a <code>data-*</code> attribute that we'll attempt to use when selecting the elements. This is handy
                when React or Vue based apps generate random class names.</small>
            </div>
          </div>
        </div>
        <div class="settings-block">
          <h4 class="settings-block-title">
            Code Generator settings
          </h4>
          <div class="settings-block-main">
            <div class="settings-group">
              <label>
                <input id="options-code-wrapAsync" type="checkbox" v-model="options.code.wrapAsync" @change="save">
                wrap code in async function
              </label>
            </div>
            <div class="settings-group">
              <label>
                <input id="options-code-headless" type="checkbox" v-model="options.code.headless" @change="save">
                set <code>headless</code> in Playwright launch options
              </label>
            </div>
            <div class="settings-group">
              <label>
                <input id="options-code-waitForNavigation" type="checkbox" v-model="options.code.waitForNavigation" @change="save">
                add <code>waitForNavigation</code> lines on navigation
              </label>
            </div>
            <div class="settings-group">
              <label>
                <input id="options-code-waitForSelectorOnClick" type="checkbox" v-model="options.code.waitForSelectorOnClick" @change="save">
                add <code>waitForSelector</code> lines before every <code>page.click()</code>
              </label>
            </div>
            <div class="settings-group">
              <label>
                <input id="options-code-blankLinesBetweenBlocks" type="checkbox" v-model="options.code.blankLinesBetweenBlocks" @change="save">
                add blank lines between code blocks
              </label>
            </div>
          </div>
        </div>
        <div class="settings-block">
          <h4 class="settings-block-title">
            Extension settings
          </h4>
          <div class="settings-block-main">
            <div class="settings-group">
              <label>
                <input id="options-telemetry" type="checkbox" v-model="options.extension.telemetry" @change="save">
                Allow recording of usage telemetry
              </label>
              <br>
              <small>We only record clicks for basic product development, no website content or input data. Data is never, ever shared with 3rd parties.</small>
            </div>
          </div>
        </div>
      </div>
      <div class="footer">
        sponsored by
        <a href="https://playwright.dev/" target="_blank">
          <img src="/images/mslogo.jpg" alt="Microsoft logo">
        </a>
        &amp
        <a href="https://checklyhq.com" target="_blank">
          <img src="/images/text_racoon_logo.svg" alt="">
        </a>
      </div>
    </div>
  </div>
</template>

<script>
  import { defaults as code } from '../../code-generator/CodeGenerator'

  const defaults = {
    code,
    extension: {
      telemetry: true
    }
  }

  export default {
    name: 'App',
    data () {
      return {
        loading: true,
        saving: false,
        options: defaults
      }
    },
    mounted () {
      this.load()
    },
    methods: {
      save () {
        this.saving = true
        this.$chrome.storage.local.set({ options: this.options }, () => {
          console.debug('saved options')
          setTimeout(() => {
            this.saving = false
          }, 500)
        })
      },
      load () {
        this.$chrome.storage.local.get('options', ({ options }) => {
          if (options) {
            console.debug('loaded options', JSON.stringify(options))
            this.options = options
          }
          this.loading = false
        })
      }
    }
  }
</script>

<style lang="scss" scoped>
  @import "~styles/_variables.scss";
  @import "~styles/_mixins.scss";

  .options {

    height: 100%;
    min-height: 580px;
    background: $gray-lighter;
    display: flex;
    flex-direction: column;
    width: 100%;
    overflow: auto;
    position: fixed;
    left: 0;
    top: 0;

    .container {
      padding: 0 2 * $spacer;
      width: 550px;
      margin: 0 auto;

      .content {
        background: white;
        padding: 2 * $spacer;
        border-radius: 4px;
        min-height: 500px;
      }

      .footer {
        @include footer();
        background: $gray-lighter;
        font-weight: normal;
        justify-content: center;
        img {
          margin-left: 8px;
          width: 80px;
          vertical-align: middle;
        }
      }

      .header {
        @include header();
        background: $gray-lighter;
        justify-content: space-between;
      }

      .settings-block {

        .settings-label {
          display: block;
          text-transform: uppercase;
          font-size: .75rem;
          font-weight: 500;
          margin-bottom: $spacer;
        }

        .settings-block-title {
          margin: 0;
          padding-bottom: $spacer;
          border-bottom: 1px solid $gray-light;
        }
        .settings-block-main {
          padding: $spacer 0;
          margin-bottom: $spacer;

          .settings-group {
            margin-bottom: $spacer;
            display: block;
          }
        }
        input[type="text"] {
          margin-bottom: 10px;
          width: 100%;
          border: 1px solid $gray-light;
          padding-left: 15px;
          height: 38px;
          font-size: 14px;
          border-radius: 10px;
          -webkit-box-sizing: border-box;
        }
      }
    }
  }
</style>
