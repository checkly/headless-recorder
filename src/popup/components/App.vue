<template>
  <div id="puppeteer-recorder" class="recorder">
    <div class="header">
      <a href="#" @click="goHome">
        Playwright recorder <span class="text-muted"><small>{{version}}</small></span>
      </a>
      <div class="left">
        <div class="recording-badge" v-show="isRecording">
          <span class="red-dot"></span>
          {{recordingBadgeText}}
        </div>
        <a href="#" @click="toggleShowHelp" class="header-button">
          <img src="/images/help.svg" alt="help" width="18px">
        </a>
        <a href="#" @click="openOptions" class="header-button">
          <img src="/images/settings.svg" alt="settings" width="18px">
        </a>
      </div>
    </div>
    <div class="main">
      <div class="tabs" v-show="!showHelp">
        <RecordingTab :code="code" :is-recording="isRecording" :live-events="liveEvents" v-show="!showResultsTab"/>
        <div class="recording-footer" v-show="!showResultsTab">
          <button class="btn btn-sm" @click="toggleRecord" :class="isRecording ? 'btn-danger' : 'btn-primary'">
            {{recordButtonText}}
          </button>
          <button class="btn btn-sm btn-primary btn-outline-primary" @click="togglePause" v-show="isRecording">
            {{pauseButtonText}}
          </button>
          <a href="#" @click="showResultsTab = true" v-show="code">view code</a>
          <checkly-badge v-if="!isRecording"></checkly-badge>
        </div>
        <ResultsTab :code="code" :copy-link-text="copyLinkText" :restart="restart" :set-copying="setCopying" v-show="showResultsTab"/>
        <div class="results-footer" v-show="showResultsTab">
          <button class="btn btn-sm btn-primary" @click="restart" v-show="code">Restart</button>
          <a href="#" v-clipboard:copy='code' @click="setCopying" v-show="code">{{copyLinkText}}</a>
        </div>
      </div>
      <HelpTab v-show="showHelp"></HelpTab>
    </div>
  </div>
</template>

<script>
  import { version } from '../../../package.json'
  import CodeGenerator from '../../code-generator/CodeGenerator'
  import RecordingTab from './RecordingTab.vue'
  import ResultsTab from './ResultsTab.vue'
  import HelpTab from './HelpTab.vue'
  import ChecklyBadge from './ChecklyBadge.vue'

  import actions from '../../models/extension-ui-actions'

export default {
    name: 'App',
    components: { ResultsTab, RecordingTab, HelpTab, ChecklyBadge },
    data () {
      return {
        code: '',
        options: {},
        showResultsTab: false,
        showHelp: false,
        liveEvents: [],
        recording: [],
        isRecording: false,
        isPaused: false,
        isCopying: false,
        bus: null,
        version
      }
    },
    mounted () {
      this.loadState(() => {
        this.trackPageView()
        if (this.isRecording) {
          console.debug('opened in recording state, fetching recording events')
          this.$chrome.storage.local.get(['recording', 'options'], ({ recording }) => {
            console.debug('loaded recording', recording)
            this.liveEvents = recording
          })
        }

        if (!this.isRecording && this.code) {
          this.showResultsTab = true
        }
      })
      this.bus = this.$chrome.extension.connect({ name: 'recordControls' })
    },
    methods: {
      toggleRecord () {
        if (this.isRecording) {
          this.stop()
        } else {
          this.start()
        }
        this.isRecording = !this.isRecording
        this.storeState()
      },
      togglePause () {
        if (this.isPaused) {
          this.bus.postMessage({ action: actions.UN_PAUSE })
          this.isPaused = false
        } else {
          this.bus.postMessage({ action: actions.PAUSE })
          this.isPaused = true
        }
        this.storeState()
      },
      start () {
        this.trackEvent('Start')
        this.cleanUp()
        console.debug('start recorder')
        this.bus.postMessage({ action: actions.START })
      },
      stop () {
        this.trackEvent('Stop')
        console.debug('stop recorder')
        this.bus.postMessage({ action: actions.STOP })

        this.$chrome.storage.local.get(['recording', 'options'], ({ recording, options }) => {
          console.debug('loaded recording', recording)
          console.debug('loaded options', options)

          this.recording = recording
          const codeOptions = options ? options.code : {}

          const codeGen = new CodeGenerator(codeOptions)
          this.code = codeGen.generate(this.recording)
          this.showResultsTab = true
          this.storeState()
        })
      },
      restart () {
        console.log('restart')
        this.cleanUp()
        this.bus.postMessage({ action: actions.CLEAN_UP })
      },
      cleanUp () {
        this.recording = this.liveEvents = []
        this.code = ''
        this.showResultsTab = this.isRecording = this.isPaused = false
        this.storeState()
      },
      openOptions () {
        this.trackEvent('Options')
        if (this.$chrome.runtime.openOptionsPage) {
          this.$chrome.runtime.openOptionsPage()
        }
      },
      loadState (cb) {
        this.$chrome.storage.local.get(['controls', 'code', 'options'], ({ controls, code, options }) => {
          if (controls) {
            this.isRecording = controls.isRecording
            this.isPaused = controls.isPaused
          }

          if (code) {
            this.code = code
          }

          if (options) {
            this.options = options
          }
          cb()
        })
      },
      storeState () {
        this.$chrome.storage.local.set({
          code: this.code,
          controls: {
            isRecording: this.isRecording,
            isPaused: this.isPaused
          }
        })
      },
      setCopying () {
        this.trackEvent('Copy')
        this.isCopying = true
        setTimeout(() => { this.isCopying = false }, 1500)
      },
      goHome () {
        this.showResultsTab = false
        this.showHelp = false
      },
      toggleShowHelp () {
        this.trackEvent('Help')
        this.showHelp = !this.showHelp
      },
      trackEvent (event) {
        if (this.options && this.options.extension && this.options.extension.telemetry) {
          if (window._gaq) window._gaq.push(['_trackEvent', event, 'clicked'])
        }
      },
      trackPageView () {
        if (this.options && this.options.extension && this.options.extension.telemetry) {
          if (window._gaq) window._gaq.push(['_trackPageview'])
        }
      }
    },
    computed: {
      recordingBadgeText () {
        return this.isPaused ? 'paused' : 'recording'
      },
      recordButtonText () {
        return this.isRecording ? 'Stop' : 'Record'
      },
      pauseButtonText () {
        return this.isPaused ? 'Resume' : 'Pause'
      },
      copyLinkText () {
        return this.isCopying ? 'copied!' : 'copy to clipboard'
      }
    }
}
</script>

<style lang="scss" scoped>
  @import "~styles/_animations.scss";
  @import "~styles/_variables.scss";
  @import "~styles/_mixins.scss";
  .recorder {
    font-size: 14px;

    .header {
      @include header();

      a {
        color: $gray-dark;
      }

      .left {
        margin-left: auto;
        display: flex;
        justify-content: flex-start;
        align-items: center;

        .recording-badge {
          color: $brand-danger;
          .red-dot {
            height: 9px;
            width: 9px;
            background-color: $brand-danger;
            border-radius: 50%;
            display: inline-block;
            margin-right: .4rem;
            vertical-align: middle;
            position: relative;
          }
        }

        .header-button {
          margin-left: $spacer;
          img {
            vertical-align: middle;
          }
        }
      }
    }

    .recording-footer {
      @include footer();
      img {
        margin-left: 8px;
        width: 80px;
        vertical-align: middle;
      }
    }
    .results-footer {
      @include footer()
    }
  }
</style>
