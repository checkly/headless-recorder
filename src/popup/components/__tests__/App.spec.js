import { createLocalVue, shallowMount } from '@vue/test-utils'
import App from '../App.vue'

const chrome = {
  storage: {
    local: {
      get: jest.fn()
    }
  },
  extension: {
    connect: jest.fn()
  }
}

const mocks = { $chrome: chrome }

describe('App.vue', () => {
  let localVue
  beforeEach(() => {
    localVue = createLocalVue()
    localVue.directive('highlightjs', () => {})
    localVue.directive('clipboard', () => {})
  })

  test('it has the correct pristine / empty state', () => {
    const wrapper = shallowMount(App, { mocks, localVue })
    expect(wrapper.element).toMatchSnapshot()
  })
})
