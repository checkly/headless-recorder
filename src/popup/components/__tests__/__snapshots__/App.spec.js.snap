// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App.vue it has the correct pristine / empty state 1`] = `
<div
  class="recorder"
  id="headless-recorder"
>
  <div
    class="header"
  >
    <a
      href="#"
    >
      
      Headless recorder 
      <span
        class="text-muted"
      >
        <small>
          0.8.1
        </small>
      </span>
    </a>
     
    <div
      class="left"
    >
      <div
        class="recording-badge"
        style="display: none;"
      >
        <span
          class="red-dot"
        />
        
        recording
      
      </div>
       
      <a
        class="header-button"
        href="#"
      >
        <img
          alt="help"
          src="/images/help.svg"
          width="18px"
        />
      </a>
       
      <a
        class="header-button"
        href="#"
      >
        <img
          alt="settings"
          src="/images/settings.svg"
          width="18px"
        />
      </a>
    </div>
  </div>
   
  <div
    class="main"
  >
    <div
      class="tabs"
    >
      <recordingtab-stub
        code=""
        liveevents=""
      />
       
      <div
        class="recording-footer"
      >
        <button
          class="btn btn-sm btn-primary"
        >
          
          Record
        
        </button>
         
        <button
          class="btn btn-sm btn-primary btn-outline-primary"
          style="display: none;"
        >
          
          Pause
        
        </button>
         
        <a
          href="#"
          style="display: none;"
        >
          view code
        </a>
         
        <checklybadge-stub />
      </div>
       
      <!---->
       
      <div
        class="results-footer"
        style="display: none;"
      >
        <button
          class="btn btn-sm btn-primary"
          style="display: none;"
        >
          Restart
        </button>
         
        <a
          href="#"
          style="display: none;"
        >
          copy to clipboard
        </a>
      </div>
    </div>
     
    <helptab-stub
      style="display: none;"
    />
  </div>
</div>
`;
